### PR TITLE
Ex 02

### PR DESCRIPTION
**Pull request by MilliamM**
Exercise 02 by Daniel Milliam Müller

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -1,8 +1,8 @@
 INF200 Exersices:
```
Comment (danielambrosius) [inline README.rst]: don't redo this. It clutters the pull request and the changes from 00 will remain when you merge back into master.

File: src/milliam_muller_ex/ex02/bubble_sort.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+    :return:
+    sorted_list: list
+        A new list of integers from input 'data' in ascending order.
+    """
+    sorted_list = list(data_list)
+    n = len(sorted_list)
+
+    i = 0
+    while i < n - 1:
```
Comment (danielambrosius) [inline src/milliam_muller_ex/ex02/bubble_sort.py]: consider changing to a for-loop using the `range()`object as iterator

File: src/milliam_muller_ex/ex02/file_stats.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+    The function converts each symbol to it's corresponding character code.
+
+    :return:
+    code_list: list
+        A list containing the frequency of each character code in the file.
+        code_list[i] gives the number of occurrences of character code 'i'.
+    """
+
+    file_name = open(text_filename, 'r', encoding='utf-8').read()
```
Comment (danielambrosius) [inline src/milliam_muller_ex/ex02/file_stats.py]: No way of knowing weather the file is closed when you do the `read() ` in a oneliner. Consider either assigning a variable to the file, then calling `.read()` on that, and dubsequently closing it. Or the preferred way is using a `with ` block.

File: src/milliam_muller_ex/ex02/file_stats.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    entry = []
+    code_list = []
+    for item in file_name:
+        entry.append(ord(item))
+
+    for n in range(len(entry)):
+        value = entry.count(n)
+        code_list.append(value)
```
Comment (danielambrosius) [inline src/milliam_muller_ex/ex02/file_stats.py]: This is somewhat of a cumbersome way of achieving this. Have a look at the sample solution.

File: src/milliam_muller_ex/ex02/message_entropy.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+    h_entropy: float
+    """
+    h_entropy = 0
+    p_n = len(message)
+    if p_n == 0:
+        return h_entropy
+    for character_i in range(256):
+        p_x = message.count(chr(character_i)) / p_n
+        if p_x > 0:
+            h_entropy += - p_x * math.log(p_x, 2)
```
Comment (danielambrosius) [inline src/milliam_muller_ex/ex02/message_entropy.py]: you should use `log2()` here.

**Comment by danielambrosius** [comment]
Also there is no author and encoding info in the scripts it looks like. This is mandatory to include from ex_03 onward

File: README.rst (hunk 1 above)
Comment (MilliamM) [reply README.rst]: Thank you for the remarks. I have been a little confused about the process of pushing to GitHub and how to properly make new branches. I'll be in the TA class on Friday to clear this up. I'm hoping I'll understand it a bit better after that. :-)

**Pull request merged**